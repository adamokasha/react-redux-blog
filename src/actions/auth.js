import axios from "axios";

export const login = (displayName, id, role) => ({
  type: "LOGIN",
  displayName,
  id,
  role
});

export const startSignup = (displayName, email, password) => {
  return async dispatch => {
    try {
      const response = await axios.post(`${process.env.API_URI}/users`, {
        displayName,
        email,
        password
      });

      const user = {
        displayName: response.data.displayName,
        id: response.data._id,
        token: response.headers["x-auth"],
        role: response.data.role
      };
      localStorage.setItem("auth", JSON.stringify(user));

      dispatch(
        login(response.data.displayName, response.data._id, response.data.role)
      );
    } catch (e) {
      throw e.response.data.errors;
    }
  };
};

export const startLogin = (email, password) => {
  return async dispatch => {
    try {
      const response = await axios.post(`${process.env.API_URI}/users/login`, {
        email,
        password
      });

      const user = {
        displayName: response.data.displayName,
        id: response.data._id,
        token: response.headers["x-auth"],
        role: response.data.role
      };
      localStorage.setItem("auth", JSON.stringify(user));

      dispatch(
        login(response.data.displayName, response.data._id, response.data.role)
      );
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const startLogout = () => {
  return async dispatch => {
    try {
      const { token } = JSON.parse(localStorage.getItem("auth"));
      localStorage.removeItem("auth");
      await axios.delete(`${process.env.API_URI}/users/me/token`, {
        headers: {
          "x-auth": token
        }
      });
      dispatch(logout());
    } catch (e) {
      localStorage.removeItem("auth");
      dispatch(logout());
    }
  };
};

export const logout = () => ({
  type: "LOGOUT"
});
