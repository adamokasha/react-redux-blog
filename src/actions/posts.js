import axios from "axios";

export const setPosts = posts => ({
  type: "SET_POSTS",
  posts
});

export const startSetPosts = () => {
  return async dispatch => {
    try {
      const response = await axios.get(
        `${process.env.REACT_APP_API_URL}/posts`
      );
      dispatch(setPosts(response.data.posts));
    } catch (e) {
      throw new Error("Could not retrieve posts...");
    }
  };
};

export const addPost = post => ({
  type: "ADD_POST",
  post
});

export const startAddPost = post => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts`,
        method: "post",
        headers: {
          "x-auth": token
        },
        data: post
      });

      dispatch(addPost(response.data.post));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const editPost = updates => ({
  type: "EDIT_POST",
  updates
});

export const startEditPost = (post, id) => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}`,
        method: "patch",
        headers: {
          "x-auth": token
        },
        data: post
      });

      dispatch(editPost(response.data.post));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const deletePost = id => ({
  type: "DELETE_POST",
  id
});

export const startDeletePost = id => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}/`,
        method: "delete",
        headers: {
          "x-auth": token
        }
      });

      dispatch(deletePost(id));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const addComment = (id, comment) => ({
  type: "ADD_COMMENT",
  id,
  comment
});

export const startAddComment = (id, comment) => {
  return async dispatch => {
    try {
      const { token } = JSON.parse(localStorage.getItem("auth"));
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}/comments`,
        method: "post",
        headers: {
          "x-auth": token
        },
        data: {
          comment
        }
      });

      dispatch(addComment(id, response.data));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};
