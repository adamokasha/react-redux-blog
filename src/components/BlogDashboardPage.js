import React from "react";
import { connect } from "react-redux";

import Header from "./Header";
import PostsList from "./PostsList";

export const BlogDashboardPage = ({ displayName }) => {
  function renderWelcome() {
    if (displayName) {
      return <h3>Welcome {displayName}.</h3>;
    }
    return;
  }
  return (
    <div>
      <Header />
      <div id="landing">
        <div className="primary-overlay">
          <div className="container h-100">
            <div className="row hero-text">
              <div className="col-lg-7">
                <h1 className="text-white mb-5">
                  React Blog - Responsive blog styled with Bootstrap 4
                </h1>
                <a
                  href="https://github.com/adamokasha/react-redux-blog"
                  target="_blank"
                  className="text-uppercase text-white hero-link"
                >
                  <u>Learn More</u>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <br />
      <div className="container mb-3">{renderWelcome()}</div>
      <div>
        <PostsList />
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  displayName: state.auth.displayName
});

export default connect(mapStateToProps)(BlogDashboardPage);
