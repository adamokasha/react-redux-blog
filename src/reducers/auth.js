const defaultAuthState = () => {
  if (localStorage.getItem("auth")) {
    return JSON.parse(localStorage.getItem("auth"));
  } else {
    return {};
  }
};

const authReducer = (state = defaultAuthState(), action) => {
  switch (action.type) {
    case "LOGIN":
      return {
        displayName: action.displayName,
        id: action.id,
        role: action.role
      };
    case "LOGOUT":
      return {};
    default:
      return state;
  }
};

export default authReducer;
