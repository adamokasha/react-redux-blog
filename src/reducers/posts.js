const postsReducer = (state = [], action) => {
  switch (action.type) {
    case "SET_POSTS":
      return action.posts;
    case "ADD_POST":
      return [...state, action.post];
    case "EDIT_POST":
      return state.map(post => {
        if (post._id === action.updates._id) {
          return { ...post, ...action.updates };
        } else {
          return post;
        }
      });
    case "ADD_COMMENT":
      return state.map(post => {
        if (post._id === action.id) {
          return {
            ...post,
            comments: [...post.comments, action.comment]
          };
        } else {
          return post;
        }
      });
    case "DELETE_POST":
      return state.filter(post => {
        if (post._id !== action.id) {
          return post;
        }
      });
    default:
      return state;
  }
};

export default postsReducer;
