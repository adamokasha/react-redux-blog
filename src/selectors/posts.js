export const sortPostsByDate = postsArray => {
  const newPostsArray = [].concat(postsArray);
  return newPostsArray.sort((a, b) => {
    if (a.createdAt > b.createdAt) {
      return -1;
    }
    if (a.createdAt < b.createdAt) {
      return 1;
    }
  });
};

export const sortCommentsByDate = commentsArray => {
  const newCommentsArray = [].concat(commentsArray);
  return newCommentsArray.sort((a, b) => {
    if (a.createdAt > b.createdAt) {
      return -1;
    }
    if (a.createdAt < b.createdAt) {
      return 1;
    }
  });
};
