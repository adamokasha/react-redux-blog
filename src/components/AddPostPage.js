import React from "react";
import { connect } from "react-redux";

import PostForm from "./PostForm";
import { startAddPost } from "../actions/posts";

export class AddPostPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: ""
    };
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  renderError = () => {
    if (this.state.error) {
      return (
        <div className="error-msg alert alert-danger" role="alert">
          <p>{this.state.error}</p>
        </div>
      );
    }
  };
  onSubmit = async post => {
    try {
      await this.props.startAddPost(post);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  render() {
    return (
      <div className="container">
        <h3 className="text-center mt-4">Add Post</h3>
        {this.renderError()}
        <PostForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post))
});

export default connect(
  null,
  mapDispatchToProps
)(AddPostPage);
