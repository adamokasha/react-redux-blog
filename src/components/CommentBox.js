import React from "react";
import { connect } from "react-redux";
import moment from "moment";

import CommentForm from "./CommentForm";
import { sortCommentsByDate } from "../selectors/posts";

export class CommentBox extends React.Component {
  renderCommentForm = () => {
    if (this.props.auth.id) {
      return <CommentForm id={this.props.id} />;
    }
    return <p className="font-italic">You must be signed in to comment...</p>;
  };
  renderComments = () => {
    if (this.props.comments.length > 0) {
      return sortCommentsByDate(this.props.comments).map(comment => {
        return (
          <li className="list-group-item pl-0" key={comment._id}>
            <p className="text-muted mb-1">
              Posted by {comment.createdBy} on{" "}
              {moment(comment.date).format("MM-DD-YY [at] HH:mm")}
            </p>
            <p className="mb-1">{comment.comment}</p>
          </li>
        );
      });
    } else {
      return <p>Be the first to comment...</p>;
    }
  };
  render() {
    return (
      <div>
        <h4>Comments</h4>
        {this.renderCommentForm()}
        <ul className="list-group-flush pl-0">{this.renderComments()}</ul>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(mapStateToProps)(CommentBox);
