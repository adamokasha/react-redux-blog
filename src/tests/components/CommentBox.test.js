import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { CommentBox } from "../../components/CommentBox";
import { withUserRole } from "../fixtures/auth";
import posts from "../fixtures/posts";

configure({ adapter: new Adapter() });

test("should render CommentBox correctly", () => {
  const wrapper = shallow(<CommentBox auth={{}} comments={[]} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render CommentBox with comments", () => {
  const wrapper = shallow(
    <CommentBox auth={{}} comments={posts[0].comments} />
  );
  expect(wrapper.find("ul li").length).toBe(2);
});

test("should render CommentForm if user logged in", () => {
  const wrapper = shallow(
    <CommentBox id={posts[0]._id} auth={withUserRole} comments={[]} />
  );
  expect(wrapper).toMatchSnapshot();
});
