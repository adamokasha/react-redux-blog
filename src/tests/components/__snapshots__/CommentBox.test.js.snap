// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CommentBox correctly 1`] = `
<div>
  <h4>
    Comments
  </h4>
  <p
    className="font-italic"
  >
    You must be signed in to comment...
  </p>
  <ul
    className="list-group-flush pl-0"
  >
    <p>
      Be the first to comment...
    </p>
  </ul>
</div>
`;

exports[`should render CommentForm if user logged in 1`] = `
<div>
  <h4>
    Comments
  </h4>
  <Connect(CommentForm)
    id="5b62cb86efcae617d8a0bef9"
  />
  <ul
    className="list-group-flush pl-0"
  >
    <p>
      Be the first to comment...
    </p>
  </ul>
</div>
`;
